common/elasticsearch/esql/esql.go

```diff
@@ -0,0 +1,164 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package esql
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/xwb1989/sqlparser"
+)
+
+// Replace ...
+// esql use replace function to apply user colName or column value replacing policy
+type Replace func(string) (string, error)
+
+// Filter ...
+// esql use filter function decide whether the policy will be applied to the column
+// only accept column names that filter(colName) == true
+type Filter func(string) bool
+
+// ESql ...
+// ESql is used to hold necessary information that required in parsing
+type ESql struct {
+	filterReplace Filter  // select the column we want to replace name
+	filterProcess Filter  // select the column we want to process value
+	replace       Replace // if selected by filterReplace, change the column name
+	process       Replace // if selected by filterProcess, change the column value
+	cadence       bool
+	pageSize      int
+	bucketNumber  int
+}
+
+// SetDefault ...
+// all members goes to default
+// should not be called if there is potential race condition
+func (e *ESql) SetDefault() {
+	e.pageSize = DefaultPageSize
+	e.bucketNumber = DefaultBucketNumber
+	e.cadence = false
+	e.replace = nil
+	e.process = nil
+	e.filterReplace = nil
+	e.filterProcess = nil
+}
+
+// NewESql ... return a new default ESql
+func NewESql() *ESql {
+	return &ESql{
+		pageSize:      DefaultPageSize,
+		bucketNumber:  DefaultBucketNumber,
+		cadence:       false,
+		process:       nil,
+		replace:       nil,
+		filterReplace: nil,
+		filterProcess: nil,
+	}
+}
+
+// SetReplace ... set up user specified column name replacement policy
+// should not be called if there is potential race condition
+func (e *ESql) SetReplace(filterArg Filter, replaceArg Replace) {
+	e.filterReplace = filterArg
+	e.replace = replaceArg
+}
+
+// SetProcess ... set up user specified column value processing policy
+// should not be called if there is potential race condition
+func (e *ESql) SetProcess(filterArg Filter, processArg Replace) {
+	e.filterProcess = filterArg
+	e.process = processArg
+}
+
+// SetPageSize ... set the number of documents returned in a non-aggregation query
+// should not be called if there is potential race condition
+func (e *ESql) SetPageSize(pageSizeArg int) {
+	e.pageSize = pageSizeArg
+}
+
+// SetBucketNum ... set the number of bucket returned in an aggregation query
+// should not be called if there is potential race condition
+func (e *ESql) SetBucketNum(bucketNumArg int) {
+	e.bucketNumber = bucketNumArg
+}
+
+// ConvertPretty ...
+// Transform sql to elasticsearch dsl, and prettify the output json
+//
+// usage:
+//  - dsl, sortField, err := e.ConvertPretty(sql, pageParam1, pageParam2, ...)
+//
+// arguments:
+//  - sql: the sql query needs conversion in string format
+//  - pagination: variadic arguments that indicates es search_after for pagination
```

Comment:
Need more description and example, otherwise users will have no idea how to use pagination. 
Also does it mean we only supporting search_after?

Reply:
If you mean scroll then yes, now only search after is supported. Let's discuss it later.